test-data/unit/check-based-union-join.test

```diff
@@ -0,0 +1,62 @@
+[case testSimpleUnionAtJoin]
+a = [1, ""]
+reveal_type(a)  # N: Revealed type is "list[int | str]"
+
+
+[case testSubtype]
+class A: pass
+class B(A): pass
+
+reveal_type([A(), B()])  # N: Revealed type is "list[__main__.A]"
+
+
+[case testConditionalExpression]
+class A: pass
+class B: pass
+
+res = A() if bool() else B()
+
+reveal_type(res)  # N: Revealed type is "__main__.A | __main__.B"
+
+
+[case testTupleUnion]
+reveal_type([(1, ""), (None, False)])  # N: Revealed type is "list[(int, str) | (None, bool)]"
+[builtins fixtures/tuple.pyi]
+
+
+[case testUnionType]
+reveal_type(type([1, ""][0]))  # N: Revealed type is "type[int] | type[str]"
+
+
+[case testAbstractTypeCollection]
+from typing import Protocol
+class Base(Protocol):
+    a: int
+class A(Base):
+    pass
+class B(Base):
+    pass
+class C(Base):
+    pass
+
+reveal_type([A, B, C])  # N: Revealed type is "list[def () -> __main__.A | def () -> __main__.B | def () -> __main__.C]"
+
+
+[case testTypeInIf]
+from typing import Type
+
+class Ba: pass
+class A(Ba): pass
+class B(Ba): pass
+class C(Ba): pass
+class D(Ba): pass
+
+def f() :
+    if 1 + 1:
+        cls: Type[Ba] = A
+    elif 1 + 1:
+        cls = B
+    elif 1 + 1:
+        cls = C
+    else:
+        cls = D
```

Comment:
what is this testing? should there be a negative test

Reply:
screw it, this is fine